PV260_Project/BusinessLayer/Services/Implementation/FileUrlService.cs

```diff
@@ -0,0 +1,61 @@
+﻿using DataLayer;
+using DataLayer.Models;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BusinessLayer.Services.Implementation
+{
+    public class FileUrlService : IFileUrlService
+    {
+        private readonly StockSystemDbContext _context;
+
+        public FileUrlService(StockSystemDbContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<FileUrl?> GetLatest()
+        {
+            return await _context.FileUrls.Where(fileUrl => fileUrl.ValidTo == null).FirstOrDefaultAsync();
+        }
+
+        public async Task<FileUrl> SetNewFileUrl(string url)

```

Comment:
Just a tip: I would implement this function like this
```
await _context.FileUrls.Where(fileUrl => fileUrl.ValidTo == null).ForEachAsync(x => x.ValidTo = DateTime.Now);
var result = await _context.FileUrls.AddAsync(new FileUrl() { Url = url });
await _context.SaveChangesAsync();
return result.Entity;
```
There is no need for transactions, because EF handles it so if something fails before saving changes, it will be reverted. Also maybe I would not return the entity directly, because if someone changes the result of this function, it changes the entity after next `SaveChanges`. There is a hack to this used in other services called `AsNoTracking()`.

Reply:
I removed the transaction. I used `AsNoTracking()` in the other query, however, it is not available for the `AddAsync` function. I did not find any way to make it work even looking at the methods on the returned entity (if I could detach it from the context). So I am going to leave it as is, if anyone knows how this could be done in EF, feel free to change.

Reply:
might have found it :D